### ISSUE TITLE
Erroneous singular matrix in .solve?

### ISSUE THREAD

**Issue by bright-star** [AtheMathmo/rulinalg#74]
I'm verifying the results of the FD port I have written using rulinalg, which consists of the following two files:
- https://gitlab.com/bright-star/electro/blob/master/examples/empty_rect.rs
- https://gitlab.com/bright-star/electro/blob/master/src/lib.rs

Running that example `cargo run --example empty_rect` with nightly gives an LUP-related error:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: 
Error { kind: DivByZero, error: StringError("Singular matrix found in LUP decomposition. A value in the diagonal of U == 0.0.") }', ../src/libcore/result.rs:799
```

It also prints out the matrix and vector it was trying to solve with, which I verified against the MATLAB source I am porting with the same params: https://github.com/bright-star/computational-em/blob/master/finite-difference/e1_uniform.m So I know that this a solvable system.

For my sanity's sake: these are identical matrices, right?

```
-4   1   0   1   0   0   0   0   0
 1  -4   1   0   1   0   0   0   0
 0   1  -4   0   0   1   0   0   0
 1   0   0  -4   1   0   1   0   0
 0   1   0   1  -4   1   0   1   0
 0   0   1   0   1  -4   0   0   1
 0   0   0   1   0   0  -4   1   0
 0   0   0   0   1   0   1  -4   1
 0   0   0   0   0   1   0   1  -4

-4   1   0   1   0   0   0   0   0
 1  -4   1   0   1   0   0   0   0
 0   1  -4   0   0   1   0   0   0
 1   0   0  -4   1   0   1   0   0
 0   1   0   1  -4   1   0   1   0
 0   0   1   0   1  -4   0   0   1
 0   0   0   1   0   0  -4   1   0
 0   0   0   0   1   0   1  -4   1
 0   0   0   0   0   1   0   1  -4
```

**Comment by AtheMathmo** [comment]
These are indeed the same matrix...

Could you give the vector too? It's possible this gives an opportunity to improve the stability of either LUP or the solver.

**Comment by bright-star** [comment]
The vector is:

```
-100
0
0
-100
0
0
-100
0
0
```

**Comment by AtheMathmo** [comment]
I spent a little time looking into this. I think there is a bug in the current implementation - the pivoting looks wrong to me (but the algorithm is working in most cases...).

I'll try to figure it out, I just wanted to let you know I'd taken a look.

**Comment by bright-star** [comment]
Thanks, I appreciate it!

**Comment by AtheMathmo** [comment]
I think I've resolved this. If you have time could you check out the PR and see if it works for you? I'm seeing the same results in rulinalg as in numpy.

**Comment by mabruckner** [comment]
I've encountered some errors with .solve too. I made a pull request that fixes my issues: #90.

**Comment by bright-star** [comment]
My issue's fixed, at least.
